File: lib/services/contact_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/contact_model.dart';

class ContactService {
  final SupabaseClient _supabase = Supabase.instance.client;

  /// Fetch paginated contacts, ordered by created_at descending
  Future<List<ContactModel>> getContacts({int page = 0, int pageSize = 20}) async {
    final start = page * pageSize;
    final end = start + pageSize - 1;

    final response = await _supabase
        .from('contacts')
        .select()
        .order('created_at', ascending: false)
        .range(start, end);

    final List<dynamic> data = response as List<dynamic>;
    return data.map((json) => ContactModel.fromJson(json)).toList();
  }

  /// Get a single contact by ID
  Future<ContactModel> getContactById(String id) async {
    final response = await _supabase
        .from('contacts')
        .select()
        .eq('id', id)
        .single();

    return ContactModel.fromJson(response);
  }

  /// Search contacts by name, email, or company
  Future<List<ContactModel>> searchContacts(String query, {int page = 0, int pageSize = 20}) async {
    final start = page * pageSize;
    final end = start + pageSize - 1;

    final response = await _supabase
        .from('contacts')
        .select()
        .or('first_name.ilike.%$query%,last_name.ilike.%$query%,email.ilike.%$query%,company_name.ilike.%$query%')
        .order('created_at', ascending: false)
        .range(start, end);

    final List<dynamic> data = response as List<dynamic>;
    return data.map((json) => ContactModel.fromJson(json)).toList();
  }

  /// Create a new contact
  Future<ContactModel> addContact(ContactModel contact) async {
    final json = contact.toJson();
    // Remove empty ID so the DB generates one via uuid_generate_v4()
    if (json['id'] == null || json['id'].toString().isEmpty) {
      json.remove('id');
    }

    final response = await _supabase
        .from('contacts')
        .insert(json)
        .select()
        .single();

    return ContactModel.fromJson(response);
  }

  /// Update an existing contact
  Future<ContactModel> updateContact(ContactModel contact) async {
    final json = contact.toJson();
    json['updated_at'] = DateTime.now().toIso8601String();

    final response = await _supabase
        .from('contacts')
        .update(json)
        .eq('id', contact.id)
        .select()
        .single();

    return ContactModel.fromJson(response);
  }

  /// Delete a contact by ID
  Future<void> deleteContact(String id) async {
    await _supabase.from('contacts').delete().eq('id', id);
  }

  /// Toggle the favorite status of a contact
  Future<void> toggleFavorite(String id, bool currentStatus) async {
    await _supabase
        .from('contacts')
        .update({
          'is_favorite': !currentStatus,
          'updated_at': DateTime.now().toIso8601String(),
        })
        .eq('id', id);
  }

  /// Get basic contact statistics
  Future<Map<String, dynamic>> getStats() async {
    final response = await _supabase.from('contacts').select('id, is_customer, is_favorite, created_at');
    final List<dynamic> data = response as List<dynamic>;

    final now = DateTime.now();
    final thirtyDaysAgo = now.subtract(const Duration(days: 30));

    int total = data.length;
    int customers = data.where((c) => c['is_customer'] == true).length;
    int leads = total - customers;
    int favorites = data.where((c) => c['is_favorite'] == true).length;
    int newThisMonth = data.where((c) {
      final createdAt = DateTime.tryParse(c['created_at'] ?? '');
      return createdAt != null && createdAt.isAfter(thirtyDaysAgo);
    }).length;

    return {
      'total': total,
      'leads': leads,
      'customers': customers,
      'favorites': favorites,
      'new_this_month': newThisMonth,
    };
  }
}

File: backend/src/services/ContactService.js
const db = require('../config/db');

/**
 * ContactService - Direct PostgreSQL CRUD operations for the contacts table.
 * Mirrors the Supabase schema_v2.sql `contacts` table structure.
 */
class ContactService {

    /**
     * Fetch paginated contacts ordered by created_at DESC
     */
    static async getContacts({ page = 0, pageSize = 20 } = {}) {
        const offset = page * pageSize;
        const result = await db.query(
            `SELECT * FROM contacts
             ORDER BY created_at DESC
             LIMIT $1 OFFSET $2`,
            [pageSize, offset]
        );
        return result.rows;
    }

    /**
     * Get single contact by ID
     */
    static async getContactById(id) {
        const result = await db.query(
            'SELECT * FROM contacts WHERE id = $1',
            [id]
        );
        if (result.rows.length === 0) {
            throw new Error('Contact not found');
        }
        return result.rows[0];
    }

    /**
     * Search contacts by name, email, or company_name
     */
    static async searchContacts(query, { page = 0, pageSize = 20 } = {}) {
        const offset = page * pageSize;
        const searchTerm = `%${query}%`;
        const result = await db.query(
            `SELECT * FROM contacts
             WHERE first_name ILIKE $1
                OR last_name ILIKE $1
                OR email ILIKE $1
                OR company_name ILIKE $1
             ORDER BY created_at DESC
             LIMIT $2 OFFSET $3`,
            [searchTerm, pageSize, offset]
        );
        return result.rows;
    }

    /**
     * Create a new contact
     */
    static async createContact(data) {
        const {
            first_name,
            last_name,
            email,
            phone,
            company_name,
            position,
            address,
            notes,
            assigned_to,
            is_customer = false,
            avatar_url,
            is_favorite = false,
        } = data;

        const result = await db.query(
            `INSERT INTO contacts
                (first_name, last_name, email, phone, company_name, position, address, notes, assigned_to, is_customer, avatar_url, is_favorite)
             VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)
             RETURNING *`,
            [first_name, last_name || '', email, phone, company_name, position, address, notes, assigned_to || null, is_customer, avatar_url, is_favorite]
        );
        return result.rows[0];
    }

    /**
     * Update an existing contact
     */
    static async updateContact(id, data) {
        const {
            first_name,
            last_name,
            email,
            phone,
            company_name,
            position,
            address,
            notes,
            assigned_to,
            is_customer,
            avatar_url,
            is_favorite,
            last_contacted,
        } = data;

        const result = await db.query(
            `UPDATE contacts SET
                first_name = COALESCE($1, first_name),
                last_name = COALESCE($2, last_name),
                email = COALESCE($3, email),
                phone = COALESCE($4, phone),
                company_name = COALESCE($5, company_name),
                position = COALESCE($6, position),
                address = COALESCE($7, address),
                notes = COALESCE($8, notes),
                assigned_to = COALESCE($9, assigned_to),
                is_customer = COALESCE($10, is_customer),
                avatar_url = COALESCE($11, avatar_url),
                is_favorite = COALESCE($12, is_favorite),
                last_contacted = COALESCE($13, last_contacted),
                updated_at = NOW()
             WHERE id = $14
             RETURNING *`,
            [first_name, last_name, email, phone, company_name, position, address, notes, assigned_to || null, is_customer, avatar_url, is_favorite, last_contacted, id]
        );

        if (result.rows.length === 0) {
            throw new Error('Contact not found');
        }
        return result.rows[0];
    }

    /**
     * Delete a contact by ID
     */
    static async deleteContact(id) {
        const result = await db.query(
            'DELETE FROM contacts WHERE id = $1 RETURNING id',
            [id]
        );
        if (result.rows.length === 0) {
            throw new Error('Contact not found');
        }
        return { deleted: true, id: result.rows[0].id };
    }

    /**
     * Toggle favorite status
     */
    static async toggleFavorite(id) {
        const result = await db.query(
            `UPDATE contacts
             SET is_favorite = NOT is_favorite, updated_at = NOW()
             WHERE id = $1
             RETURNING *`,
            [id]
        );
        if (result.rows.length === 0) {
            throw new Error('Contact not found');
        }
        return result.rows[0];
    }

    /**
     * Get contact statistics
     */
    static async getStats() {
        const result = await db.query(`
            SELECT
                COUNT(*) AS total,
                COUNT(*) FILTER (WHERE is_customer = false) AS leads,
                COUNT(*) FILTER (WHERE is_customer = true) AS customers,
                COUNT(*) FILTER (WHERE is_favorite = true) AS favorites,
                COUNT(*) FILTER (WHERE created_at >= NOW() - INTERVAL '30 days') AS new_this_month
            FROM contacts
        `);
        return result.rows[0];
    }
}

module.exports = ContactService;
